plotly/utils.py

```diff
@@ -444,3 +446,47 @@ def set_sharing_and_world_readable(option_set):
             option_set['world_readable'] = True
         else:
             option_set['world_readable'] = False
+
+
+def _default_memoize_key_function(*args, **kwargs):
+    """Factored out in case we want to allow callers to specify this func."""
+    if kwargs:
+        # frozenset is used to ensure hashability
+        return args, frozenset(kwargs.items())
+    else:
+        return args
+
+
+def memoize(maxsize=128):
+    """
+    Memoize a function by its arguments. Note, if the wrapped function returns
+    a mutable result, the caller is responsible for *not* mutating the result
+    as it will mutate the cache itself.
+
+    :param (int|None) maxsize: Limit the number of cached results. This is a
+                               simple way to prevent memory leaks. Setting this
+                               to `None` will remember *all* calls. The 128
```

Comment:
Yah, this mimics [lru_cache](https://docs.python.org/3/library/functools.html#functools.lru_cache). I'd have tried to use this, but it's not available in Python < 3.2.